### PR TITLE
refactor(lsp): consolidate the different floating window methods into `open_floating_preview`

### LINKED ISSUE #12281: LSP: How should hover on longer text be displayed?
In a lot of languages, it's going to be common that there will be longer signatures. For example, when using Scala, it's not uncommon to trigger a hover request on something that has a half dozen parameters. I came across this today for example. I triggered the following hover request:

```
[Trace - 05:35:40 PM] Sending response 'textDocument/hover - (3)'. Processing request took 19ms
Result: {
  "contents": {
    "kind": "markdown",
    "value": "```scala\ndef toMarkdown(input: Input, context: Context, inputFile: InputFile, siteVariables: Map[String,String], reporter: Reporter, settings: Settings): String\n```"
  },
  "range": {
    "start": {
      "line": 81,
      "character": 13
    },
    "end": {
      "line": 81,
      "character": 89
    }
  }
}
```

Currently, `vim.lsp.buf.hover()` will display this like the following. I was split vertically when I did this.

Screenshot 2020-05-08 at 17 35 55

In some of the other LSP Vim clients they will wrap in this scenario. For example, here is what coc.nvim does for the same hover request:

Screenshot 2020-05-08 at 17 35 14

What do you think about doing some type of determination on the viewport size and wrapping if necessary? Or there may be another better solution?

### LINKED ISSUE #14627: `vim.lsp.diagnostic.show_line_diagnostics` with `set foldmethod=indent`
### Actual behaviour
When using nvim lsp and having `set foldmethod=indent` the `vim.lsp.diagnostic.show_line_diagnostics` window has the errors folded, and since it seems it can't be focused (https://gitter.im/neovim/neovim?at=60a947e243fc4a24c5bcc372) the errors are unreadable.

### Expected behaviour
Being able to focus the `vim.lsp.diagnostic.show_line_diagnostics` window, possibly in a similar way as the `vim.lsp.buf.hover` window.

### PR DESCRIPTION
**Pull request by folke**
There were currently multiple ways to open floating windows.

This PR consolidates all the functionality under `open_floating_preview`

* `vim.lsp.util.preview_location` is now also focusable
*  `vim.lsp.diagnostics.show_line_diagnostics` is now also focusable
* added to `open_floating_preview`:
  - `opts.focus_id` to make it focusable
  - `opts.fancy_markdown` defaults to `true`
  - if `syntax == "markdown"` and `opts.fancy_markdown`, then the fancy rendering is used (old `fancy_floating_markdown`)
  - added `q` to close the float when focused
* deprecated methods:
  - `focusable_float`
  - `focusable_preview`
  - `fancy_floating_markdown`
* fixes:
  - `lsp_markdown.vim` always force using the default `markdown.vim` (so not any other installed markdown syntax)
  - for the fancy markdown, it is needed to apply the `lsp_markdown` regions seperately instead of setting it on the buffer. Otherwise, there's situations where the markdown syntax `bleeds` through the other syntax regions.
  - fancy markdown: keep track of syntaxes we already included. no need to include a syntax more than once
  - only apply `close_events` when they're not empty
  - disable folding in the popup (fixes #14627)
  - enable wrapping (fixes #12281)
* utility method `format_fancy_markdown`, that can be used by external plugins to format markdown in given buffers similar to the lsp floating windows. (for integration with compe, saga, ...)

Once the PR is approved, I'll create seperate commits for the different changes

## Where are floating windows used right now?

* `vim.lsp.util.preview_location`
* `vim.lsp.handlers.signature_help`
* `vim.lsp.handlers.hover`
* `vim.lsp.diagnostics.show_line_diagnostics`

### CODE REVIEW

File: runtime/syntax/lsp_markdown.vim
Comment (nanotee) [inline runtime/syntax/lsp_markdown.vim:8]: I think you can use `<sfile>` for this: 

```suggestion
execute 'source' expand('<sfile>:p:h') .. '/markdown.vim'
```

Not sure if appending `'/markdown.vim'` works on Windows since it uses `\` as a path separator

Comment (folke) [reply runtime/syntax/lsp_markdown.vim]: Awesome, just tested it and it all works as expected. Thanks!

I'm also not sure if this works on windows... If anyone else know, let me know!

File: runtime/lua/vim/lsp/util.lua
Comment (mjlbach) [inline runtime/lua/vim/lsp/util.lua:1096]: Not sold on the name, maybe "stylize_markdown"?

File: runtime/lua/vim/lsp/util.lua
Comment (mjlbach) [inline runtime/lua/vim/lsp/util.lua:1352]: I'm still not a fan of matching based on the syntax rule set (see earlier comment abou fancy_markdown)

File: runtime/lua/vim/lsp/util.lua
Comment (clason) [reply runtime/lua/vim/lsp/util.lua:1096]: `parse_markdown`? `apply_markdown_formatting`?

File: runtime/lua/vim/lsp/util.lua
Comment (mjlbach) [inline runtime/lua/vim/lsp/util.lua:1341]: I think there's a bug in the issue tracker where two windows can exist simultaneously from handlers, not sure if if focus_id and bufnr will currently always guarantee the correct buffer (it should)

File: runtime/lua/vim/lsp/util.lua
Comment (mjlbach) [inline runtime/lua/vim/lsp/util.lua:1346]: Do we not already have bufnr available here from 1330?

File: runtime/lua/vim/lsp/util.lua
Comment (folke) [reply runtime/lua/vim/lsp/util.lua:1096]: `stylize_markdown` works for me.

File: runtime/lua/vim/lsp/util.lua
Comment (folke) [reply runtime/lua/vim/lsp/util.lua:1352]: What do you mean exactly?

Previously, the fancy formatting was activated when the syntax was `lsp_markdown`.
I changed that to an option instead. The option only works for markdown of course.

Or would you keep a separate method for this instead of making it part of `open_floating_preview`?

What would you propose?

File: runtime/lua/vim/lsp/util.lua
Comment (folke) [reply runtime/lua/vim/lsp/util.lua:1346]: nope, this would be the bufnr of the "existing" floating window, not the current bufnr.

File: runtime/lua/vim/lsp/util.lua
Comment (folke) [reply runtime/lua/vim/lsp/util.lua:1096]: although, `format_markdown` does also make sense to me. We are literally formatting the markdown to look good in the floating window, by making small textual changes (adding separators, removing the code block delimiters) and applying the syntax highlighting.

File: runtime/lua/vim/lsp/util.lua
Comment (folke) [reply runtime/lua/vim/lsp/util.lua:1341]: I'm not 100% sure I understand what you mean, but the bufnr should be correct I think, since `focus_id` is something like `hover` or `signature`.

If you set the cursor on a method that has both signature help and hover doc. Then pressing `hover`, `signature_help`, `hover` will correctly focus the hover foat.

Comment (mjlbach) [reply runtime/lua/vim/lsp/util.lua:1341]: As in, focus_id is not actually a numerical (unique per windows) but categorical ID (unique per window type), and there is an open bug where multiple windows can be open at a time, not sure if it's possible for multiple windows with the same categorical ID to be open at a time

File: runtime/lua/vim/lsp/util.lua
Comment (mjlbach) [reply runtime/lua/vim/lsp/util.lua:1346]: Ah great :) Thanks

File: runtime/lua/vim/lsp/util.lua
Comment (mjlbach) [reply runtime/lua/vim/lsp/util.lua:1352]: I would just make it `opts.format_markdown` if true, use format_markdown, else, just display it as syntax markdown and let the gui handle it

File: runtime/lua/vim/lsp/util.lua
Comment (folke) [reply runtime/lua/vim/lsp/util.lua:1341]: that property is set on the floating window, not the parent window, so this should not be an issue?

Technically multiple floats can exists for the same focus_id, but becuase of the default close events, this won't happen.

If you'd change (or don't set) the events, then I believe it would still work as expected:
1. go to method one and `hover`
2. got to method two and `hover`
3. the open hover flloat will get the focus

The scenario above doesn't won't happen because of the default close events, but it could be changed.

File: runtime/lua/vim/lsp/util.lua
Comment (mjlbach) [reply runtime/lua/vim/lsp/util.lua:1096]: I dislike "render" because I see a lot on reddit that neovim doesn't "render" our markdown, when in reality it's the application of conceal rules and highlighting, which I don't know if ever have been historically grouped under rendering when describing neovim. Up to you, I think stylize is a superset of formatting which is why I suggested it (because we do format, but we also conceal).

Comment (folke) [reply runtime/lua/vim/lsp/util.lua:1096]: ok, sounds good. I'll use stylize

File: runtime/lua/vim/lsp/util.lua
Comment (folke) [reply runtime/lua/vim/lsp/util.lua:1352]: @mjlbach that's how it works right now no? :)

(apart from the name change to `stylize_markdown`)

Comment (mjlbach) [reply runtime/lua/vim/lsp/util.lua:1352]: `local is_fancy_markdown = syntax == "markdown" and opts.fancy_markdown`

I'm saying don't match on syntax, match on opts.stylize markdown which defaults to true for the function

Comment (folke) [reply runtime/lua/vim/lsp/util.lua:1352]: The same function is also used for non-markdown stuff, so if we don't extra check on `syntax` we would apply the special markdown formatting for other syntaxes?

I really think we need to only call stylize markdown if what we're putting in the float is effectively markdown.

Comment (folke) [reply runtime/lua/vim/lsp/util.lua:1352]: This is the function def:

image

Comment (mjlbach) [reply runtime/lua/vim/lsp/util.lua:1352]: Yeah, I guess I'm saying the function argument "syntax" seems a bit overloaded to me, as syntax "markdown"/"lsp_markdown" conditionally change a lot of paths in the code in the ways that aren't expected. I would say opts.sytlize_markdown = true + a check that markdown/lsp_markdown are the set syntax. does that make sense?

File: runtime/lua/vim/lsp/util.lua
Comment (mjlbach) [reply runtime/lua/vim/lsp/util.lua:1341]: I'm aware, but if you're searching for focus_id's across all floating windows, and multiple floating windows have the same focus_id, I believe that will cause issues.

The issue in question is https://github.com/neovim/neovim/issues/11508

File: runtime/lua/vim/lsp/util.lua
Comment (mjlbach) [reply runtime/lua/vim/lsp/util.lua:1352]: > That is effectively what we're doing right now.

It is, but the interface is different. you should be able to pass syntax="markdown" and opts.stylize_markdown = false as a gui

Comment (mjlbach) [reply runtime/lua/vim/lsp/util.lua:1352]: I don't think it's weird, we only call it 2-3 places so we can just pass in opts.stylize_markdown = true for hover, etc.

Comment (folke) [reply runtime/lua/vim/lsp/util.lua:1352]: > It is, but the interface is different. you should be able to pass syntax="markdown" and opts.stylize_markdown = false as a gui

And why wouldn't you be able to do that now?

File: runtime/lua/vim/lsp/util.lua
Comment (mjlbach) [inline runtime/lua/vim/lsp/util.lua:1370]: Because of this line, I'm saying we should only turn this on if stylize_markdown is true.

Comment (folke) [reply runtime/lua/vim/lsp/util.lua:1370]: I'm fine with changing that to something like `if syntax == "markdown" and opts.stylize_markdown then`.

Comment (mjlbach) [reply runtime/lua/vim/lsp/util.lua:1370]: sgtm, if you make the changes we discussed re naming I think we're good after that.

Comment (mjlbach) [reply runtime/lua/vim/lsp/util.lua:1370]: the focus_id thing notwithstanding, not sure what to do about that, it's probably fine in practice for now.

File: runtime/lua/vim/lsp/util.lua
Comment (folke) [reply runtime/lua/vim/lsp/util.lua:1341]: yeah, it depends on how we want the popups to behave.

Right now there can be only one popup per focus_id open at one time.

Do we want to change this to never allow more than one lsp poup open? I can change it like that.

Comment (mjlbach) [reply runtime/lua/vim/lsp/util.lua:1341]: Yeah, I think we shouldn't allow more than one popup per focus_id type for now (it's how it's supposed to be right now).

Comment (folke) [reply runtime/lua/vim/lsp/util.lua:1341]: Just to be sure, I don't make any changes then right? :)

> we shouldn't allow more than one popup per focus_id type for now

that's how it currently works. You can have multiple floats open, but at most one for every focus_id type. So never two or more `hover` floats for example.

Comment (mjlbach) [reply runtime/lua/vim/lsp/util.lua:1341]: Yes, my point in bringing it up is there is currently an open bug report that implies that there is a corner case in which we can have multiple floating windows of the same type (hover, signature help, etc.) as well as multiple of different types. If with this PR, the former behavior is fixed (or at least, not readily reproducible), we shouldn't block on merging this, I just brought it up in review as this is possibly undefined behavior.

**Comment by mjlbach** [comment]
Nice, I'll wait for review from @mfussenegger. Also, we need to test on windows. I can do that later.

**Comment by mjlbach** [comment]
Also to close https://github.com/neovim/neovim/issues/14627, should we just disable folding in this PR as well?

Maybe also close https://github.com/neovim/neovim/issues/12281 and make line wrap the default.

**Comment by neovim-discourse** [comment]
This pull request has been mentioned on **Neovim Discourse**. There might be relevant details there:

https://neovim.discourse.group/t/how-to-modify-inbuilt-line-diagnostics-popup-to-get-wrapped-lines/543/3

**Comment by folke** [comment]
> Also to close https://github.com/neovim/neovim/issues/14627, should we just disable folding in this PR as well?
> 
> Maybe also close https://github.com/neovim/neovim/issues/12281 and make line wrap the default.

Yep, makes sense to include those two as well. Small changes

File: runtime/lua/vim/lsp/util.lua
Comment (folke) [inline runtime/lua/vim/lsp/util.lua:1381]: @mjlbach see the folding and wrapping options above. If any other options would make sense, let me kow.

Comment (mjlbach) [reply runtime/lua/vim/lsp/util.lua:1381]: I think this is good? Seems to address the current open issues/complaints, we can also make more iterative changes if people give feedback.

Comment (folke) [reply runtime/lua/vim/lsp/util.lua:1381]: yep, agree 👍

Comment (clason) [reply runtime/lua/vim/lsp/util.lua:1381]: Let me test those changes; they seem to be a bit intrusive for me.
(I use `floating_preview` to preview a location in a file, and I would like to see that as close to what my normal buffer is showing me. Previously, this was done by a logic that takes `wrap` from the buffer itself while still allowing it to be overriden.)

Comment (folke) [reply runtime/lua/vim/lsp/util.lua:1381]: @clason maybe I should add an option for the wrapping?

Comment (folke) [reply runtime/lua/vim/lsp/util.lua:1381]: there's wrapping code in there already. I'll look into that first.

Comment (clason) [reply runtime/lua/vim/lsp/util.lua:1381]: previously we had
```lua
  opts.wrap_at = opts.wrap_at or (vim.wo["wrap"] and api.nvim_win_get_width(0))
```
which is then passed to a `_make_floating_popup_size` function. (Took a few iterations to get it working right, too ;))

Comment (mjlbach) [reply runtime/lua/vim/lsp/util.lua:1381]: @clason can you clarify the regression? I don't see a case in which we would want a floating window to not have wrapped text.

Comment (clason) [reply runtime/lua/vim/lsp/util.lua:1381]: as I said, I haven't tested it yet but want to to make sure (not at my dev machine atm). What struck me was not the fact that there is wrapping (although some people might not like it; I'll look at previous issues on that) but the options for styling (`breakindent`, with specific width).

Comment (folke) [reply runtime/lua/vim/lsp/util.lua:1381]: That `opts.wrap_at` code looks really odd. It also never sets actually wrapping on the floating window. It's only used to calculate the height of the popup based of the `wrap` property of the parent window. (which is probably inherited by the float?)

I'll change it so that the `opts.wrap_at` defaults to true.

Comment (clason) [reply runtime/lua/vim/lsp/util.lua:1381]: There was https://github.com/neovim/neovim/issues/12998, for example

Comment (folke) [reply runtime/lua/vim/lsp/util.lua:1381]: @clason yeah, I wasn't sure about the `breakindent`. Maybe I should leave that out for now?

Comment (clason) [reply runtime/lua/vim/lsp/util.lua:1381]: It's only used to calculate the height of the popup based of the `wrap` property of the parent window. (which is probably inherited by the float?)

Yes.

Comment (clason) [reply runtime/lua/vim/lsp/util.lua:1381]: Give me five minutes to check out the current version.

**Comment by runiq** [comment]
Thank you for picking this up

File: runtime/lua/vim/lsp/util.lua
Comment (clason) [reply runtime/lua/vim/lsp/util.lua]: Yeah, I think that's not going to be universally popular. I'd suggest not setting these options (`linebreak` and `breakindent`) so that they are -- if set -- inherited by the user's config. That seems to work well for `hover` and `preview_location` for me.

Comment (clason) [reply runtime/lua/vim/lsp/util.lua]: (And I'm even using almost the same options apart from the `breakindentopt`, and it's still bugging me ;))

Comment (folke) [reply runtime/lua/vim/lsp/util.lua]: @clason I just pushed some changes that probably make more sense with regards to `wrap`:

* `open_floating_preview` has a new `wrap` option that defaults to true
* if `wrap=true`, then we use `wrap_to` to calculate the height and width **AND** we enable wrapping on the float

Comment (clason) [reply runtime/lua/vim/lsp/util.lua]: That still sets unwanted options (which were my main point of contention, not the wrapping itself) ;)

Comment (mjlbach) [reply runtime/lua/vim/lsp/util.lua]: Philosophically I think it's ok to be opinionated on these windows and not add too many options (which increases complexity) as a lot of people are just going to copy/past and override the handler anyways.

Comment (clason) [reply runtime/lua/vim/lsp/util.lua]: To be clear: even _if_ I want wrapping in my floats (which I do), I don't want _your_ options; the float should respect the options I set in my `init.vim` (or not). Which seems to work nicely if the options are simply not set here.

Comment (folke) [reply runtime/lua/vim/lsp/util.lua]: @clason ok, I'll remove the indent stuff. 
But maybe keep the `linebreak`? (wrap around words)

Comment (folke) [reply runtime/lua/vim/lsp/util.lua]: @clason I get your point. I'll remove the indent and linebreak, but will keep the `wrap`

Comment (folke) [reply runtime/lua/vim/lsp/util.lua]: @clason done

Comment (clason) [reply runtime/lua/vim/lsp/util.lua]: Why? If people want linebreak, they can set it in their `init.vim`. If they don't, they can't easily override it.

(I disagree with the "opinionated windows" since this code is not just used to show simple `hover` text but also complete previews of parts of my buffer.)

Comment (clason) [reply runtime/lua/vim/lsp/util.lua]: @mjlbach To be clear, I'm not saying "add options and overrides", I'm saying "don't set options and let `init.vim` handle this".

Comment (mjlbach) [reply runtime/lua/vim/lsp/util.lua]: Sure, I haven't actually used code preview beyond testing it, so I'm not sure what the implications are of setting options that may/should be inherited from global neovim settings.

Comment (folke) [reply runtime/lua/vim/lsp/util.lua]: I agree. So now it will still enable `wrap` on the window, but won't set `linebreak` or `breakindent`. Those setting will now simply be based off the options the user has set in their init.{vim,lua}

**Comment by mjlbach** [comment]
Still need to test on windows, but I can't actually build on windows atm.

File: runtime/lua/vim/lsp/util.lua
Comment (clason) [reply runtime/lua/vim/lsp/util.lua]: (Sorry, was typing while you commented; didn't want to beat a dead horse. Now it's fine, thank you!)

**Review by mfussenegger — APPROVED** [review]
@mjlbach looks good to me

**Comment by runiq** [comment]
If you need a unit test, I wrote one in https://github.com/neovim/neovim/pull/12720/commits/6ab43d9eaef9e52c6b4b34b6c88e9904c76b1b81

**Comment by folke** [comment]
> If you need a unit test, I wrote one in https://github.com/neovim/neovim/pull/12720/commits/6ab43d9eaef9e52c6b4b34b6c88e9904c76b1b81 

Yeah, I saw that. I was planning to integrate it and basically forgot about it 🤦‍♂️ I'm not home right now, but I can add this tomorrow. Thanks!

**Comment by mjlbach** [comment]
@folke you can add this in a follow-up PR, I finally tested this on windows and it seems to be working as expected. I'll merge this so we have time to get feedback before we cut 0.5

**Comment by folke** [comment]
@mjlbach sounds good. Thanks for merging! 🚀

**Comment by mjlbach** [comment]
Thanks for addressing all of the review points! I think this is a big UX improvement.

**Comment by shadmansaleh** [comment]
Ok . Is there any way for me to make `vim.lsp.diagnostics.show_line_diagnostics` not focusable in my config ?

**Comment by folke** [comment]
@shadmansaleh not right now, but I can easily make it configurable as part of https://github.com/neovim/neovim/pull/14681

What would be your use-case? If you don't want to focus it, just dont hit the same keybinding twice?

**Comment by shadmansaleh** [comment]
> @shadmansaleh not right now, but I can easily make it configurable as part of #14681

I've opened a pr :)

> What would be your use-case? If you don't want to focus it, just dont hit the same keybinding twice?

For me it's bound to a `autocommand` so kind of a problem . When diagnostics float is focusable window nevigation bindings get inside those floats to , which is just annoying :P

**Comment by folke** [comment]
@shadmansaleh right, that would be annoying indeed :)

**Comment by David-Else** [comment]
@folke Seems like https://github.com/neovim/neovim/pull/14617 is the final piece for an awesome typescript LSP experience, does it need updating in relation to this PR, or can it be merged?

**Comment by folke** [comment]
@David-Else I'll rework that one to be included with https://github.com/neovim/neovim/pull/14681

**Pull request merged**
